drivers/dma/dma_iproc_pax_v2.c

```diff
@@ -0,0 +1,940 @@
+	ring->curr.sync_data.opaque = 0x0;
+	/* pkt count for active dma xfer */
+	ring->curr.sync_data.total_pkts = 0x0;
+}
+
+static int init_rm(struct dma_iproc_pax_data *pd)
+{
+	int ret = -ETIMEDOUT, timeout = 1000;
+
+	k_mutex_lock(&pd->dma_lock, K_FOREVER);
+	/* Wait for Ring Manager ready */
+	do {
+		LOG_DBG("Waiting for RM HW init\n");
+		if ((sys_read32(RM_COMM_REG(pd, RM_MAIN_HW_INIT_DONE)) &
+		    RM_MAIN_HW_INIT_DONE_MASK)) {
+			ret = 0;
+			break;
+		}
+		k_sleep(K_MSEC(1));
+	} while (--timeout);
+	k_mutex_unlock(&pd->dma_lock);
+
+	if (!timeout) {
+		LOG_WRN("RM HW Init timedout!\n");
+	} else {
+		LOG_INF("PAX DMA RM HW Init Done\n");
+	}
+
+	return ret;
+}
+
+static void rm_cfg_start(struct dma_iproc_pax_data *pd)
+{
+	uint32_t val;
+
+	k_mutex_lock(&pd->dma_lock, K_FOREVER);
+
+	/* set config done 0, enable toggle mode */
+	val = sys_read32(RM_COMM_REG(pd, RM_CONTROL));
+	val &= ~RM_CONTROL_CONFIG_DONE;
+	sys_write32(val, RM_COMM_REG(pd, RM_CONTROL));
+
+	val &= ~(RM_CONTROL_MODE_MASK << RM_CONTROL_MODE_SHIFT);
+	val |= (RM_CONTROL_MODE_ALL_BD_TOGGLE << RM_CONTROL_MODE_SHIFT);
+	sys_write32(val, RM_COMM_REG(pd, RM_CONTROL));
+	sys_write32(RM_MSI_DISABLE_MASK, RM_COMM_REG(pd, RM_MSI_DISABLE));
+
+	val = sys_read32(RM_COMM_REG(pd, RM_AXI_READ_BURST_THRESHOLD));
+	val &= ~(THRESHOLD_CFG_RD_FIFO_MAX_THRESHOLD_MASK <<
+		 THRESHOLD_CFG_RD_FIFO_MAX_THRESHOLD_SHIFT);
+	val |= THRESHOLD_CFG_RD_FIFO_MAX_THRESHOLD_SHIFT_VAL <<
+	       THRESHOLD_CFG_RD_FIFO_MAX_THRESHOLD_SHIFT;
+	sys_write32(val, RM_COMM_REG(pd, RM_AXI_READ_BURST_THRESHOLD));
+
+	val = sys_read32(RM_COMM_REG(pd, RM_FIFO_FULL_THRESHOLD));
+	val &= ~(PKT_ALIGNMENT_BD_FIFO_FULL_THRESHOLD_MASK <<
+		 PKT_ALIGNMENT_BD_FIFO_FULL_THRESHOLD_SHIFT);
+	val |= PKT_ALIGNMENT_BD_FIFO_FULL_THRESHOLD_VAL <<
+	       PKT_ALIGNMENT_BD_FIFO_FULL_THRESHOLD_SHIFT;
+
+	val &= ~(BD_FIFO_FULL_THRESHOLD_MASK <<
+		 BD_FIFO_FULL_THRESHOLD_SHIFT);
+	val |= BD_FIFO_FULL_THRESHOLD_VAL << BD_FIFO_FULL_THRESHOLD_SHIFT;
+	sys_write32(val, RM_COMM_REG(pd, RM_FIFO_FULL_THRESHOLD));
+
+	/* Enable Line interrupt */
+	val = sys_read32(RM_COMM_REG(pd, RM_CONTROL));
+	val |= RM_CONTROL_LINE_INTR_EN;
+	sys_write32(val, RM_COMM_REG(pd, RM_CONTROL));
+
+	/* Enable AE_TIMEOUT */
+	sys_write32(AE_TIMEOUT_VAL, RM_COMM_REG(pd, AE_TIMEOUT));
+	val = sys_read32(RM_COMM_REG(pd, RM_CONTROL));
+	val |= RM_CONTROL_AE_TIMEOUT_EN;
+	sys_write32(val, RM_COMM_REG(pd, RM_CONTROL));
+
+	/* AE (Acceleration Engine) grouping to group '0' */
+	val = sys_read32(RM_COMM_REG(pd, AE0_AE_CONTROL));
+	val &= ~AE_CTRL_AE_GROUP_MASK;
+	sys_write32(val, RM_COMM_REG(pd, AE0_AE_CONTROL));
+	val |= AE_CONTROL_ACTIVE;
+	sys_write32(val, RM_COMM_REG(pd, AE0_AE_CONTROL));
+
+	/* AXI read/write channel enable */
+	val = sys_read32(RM_COMM_REG(pd, RM_AXI_CONTROL));
+	val |= (RM_AXI_CONTROL_RD_CH_EN | RM_AXI_CONTROL_WR_CH_EN);
+	sys_write32(val, RM_COMM_REG(pd, RM_AXI_CONTROL));
+
+	/* Tune RM control programming for 4 rings */
```

Comment:
why 4 rings here?

Reply:
V1 RM supports 4 rings and V2 supports 8 rings.

We can set it to 8 but only 4 is used as of now.

Reply:
can you provide 8 ring settings as well?